### ISSUE TITLE
(enh) Needs passthru_combo (to support Any.KEY)

### ISSUE THREAD

**Issue by RedBearAK** [joshgoebel/keyszer#61]
Nested keys are working in general, but I've run into a situation where it would be helpful to be able to have an "any key" sort of option that leads to a specific output. As in, any key that doesn't trigger the other options defined in the nested group. This would be like a Bash `case *` option for input trees, when you want to define what happens when the user enters input that isn't part of the other defined case options like Y/N. 

```bash
    case $yn in
        [Yy]* ) make install; break;;
        [Nn]* ) exit;;
        * ) echo "Please answer yes or no.";;
    esac
```

Right now I don't know of a way to accomplish this sort of "default" output from a nested group, when non-specified key is the next key. I can only do specific keys like `Esc` or `Space`. I'd like to be able to do something like the last line here. 

```python
    C("Alt-Grave"): {                       # Dead Keys Accent Grave
        C("Esc"):                   ST("\u0060"),   # Leave accent character if dead keys Escaped
        C("Space"):                 ST("\u0060"),   # Leave accent character if dead keys Escaped
        C("A"):                     ST("\u00E0"),
        C("E"):                     ST("\u00E8"),
        C("I"):                     ST("\u00EC"),
        C("O"):                     ST("\u00F2"),
        C("U"):                     ST("\u00F9"),
        C("Shift-A"):               ST("\u00C0"),
        C("Shift-E"):               ST("\u00C8"),
        C("Shift-I"):               ST("\u00CC"),
        C("Shift-O"):               ST("\u00D2"),
        C("Shift-U"):               ST("\u00D9"),
        C("*"):                     ST("\u0060"),
    },
```

But there also needs to be a way to allow the key that was just pressed to become the output, without being specifically mapped on the right side. The way this works in the AutoHotKey version is like this: 

```autohotkey
    ; Grave accent: Option+`, then key to accent
    $!SC029::
        ; Use Apple "dead keys" Option key method to attach accents to next character typed
        ; Grave accent activated by Option+` (Alt plus scan code SC029, or !SC029)
        StringCaseSense, On
        ; watch next input string
        Input, UserInput, L1 
        Switch UserInput {
            Case Esc: Return            ; Watch for Escape key, cancel dead keys sequence
            Case "a": Send, {U+00E0}              ; à {U+00E0} (Alt+0224)
            Case "e": Send, {U+00E8}              ; è {U+00E8} (Alt+0232)
            Case "i": Send, {U+00EC}              ; ì {U+00EC} (Alt+0236)
            Case "o": Send, {U+00F2}              ; ò {U+00F2} (Alt+0242)
            Case "u": Send, {U+00F9}              ; ù {U+00F9} (Alt+0249)
            Case "A": Send, {U+00C0}              ; À {U+00C0} (Alt+0192)
            Case "E": Send, {U+00C8}              ; È {U+00C8} (Alt+0200)
            Case "I": Send, {U+00CC}              ; Ì {U+00CC} (Alt+0204)
            Case "O": Send, {U+00D2}              ; Ò {U+00D2} (Alt+0210)
            Case "U": Send, {U+00D9}              ; Ù {U+00D9} (Alt+0217)
            Default: Send, %UserInput%          ; No match? Send input through.
        }
        Return
```

The last line here with `Default` just sends the typed character through if it doesn't match any of the `Case` lines. With AHK, `Default` can actually come anywhere in the `Switch` statement, it doesn't have to be at the end.

**Comment by joshgoebel** [comment]
I understand the request, but what is the WHY behind this request?

**Comment by RedBearAK** [comment]
@joshgoebel 

> what is the WHY behind this request?

Don't really understand the question. This relates to the Option key special characters scheme. 

The behavior on macOS with "dead keys" accents is if you hit Escape or Space you cancel the dead keys sequence but it leaves the accent character, without combining it with anything. It just becomes a regular character. Before that, typing one of the valid keys combines it with that accent. If you type any other key that isn't valid for that particular accent, you wind up with the accent character and then the character you just typed (separately). I think this is their way of showing you that the key you typed couldn't have that accent applied. I'm just trying to mimic this as closely as possible. 

Right now what's happening is when I hit a "wrong" key after the initial dead keys accent shortcut, nothing happens. So basically the character gets eaten rather than coming out as output. And there is no way that I know of to send out the accent character and then whatever the input character was as an auto-adapting macro based on the input.

**Comment by joshgoebel** [comment]
Can you write out both workflows step by step/key by key?  The broken one and the working one?

**Comment by joshgoebel** [comment]
Ok, so lets take one thing at a time... it seems first we need to bring back the concept of `passthru` as in `passthru_combo`... except this time do it properly, it should call `_output.send_combo(combo)` and pass it the input_combo. (previously it only passed thru the key, not the entire combo)... it's possible long term you'd want both options but here it seems like you want to pass the input COMBO, not the input key.

You should need to:

- add to `ComboHint`
- add alias name to `config_api`
- add branch in `handle_commands` to handle `ComboHint.passthru_combo` - and I think yuo'd want to return `True` to reset the mode (esp the submap)

At that point you should be able to add a few *explicit* keys and test if that works, then secondly we could circle back to the idea of "Default/any"... think you could make an attempt at that?

**Comment by RedBearAK** [comment]
> think you could make an attempt at that?

The location of the question makes it a bit ambiguous which part you're asking me to tackle. But I'm assuming it's not the "Default/any" since you said that would be something to circle back to. 

What's being described here does at least seem logical to me as to what needs to happen internally to make the pass through feature work. But I'd estimate it's about 2-3 levels above what I've attempted to implement before with mostly isolated functions. Although I may be misjudging the difficulty due to lack of experience. 

> I do agree that both these things seem in scope and reasonable to have.

Glad to have that acknowledgement that the desired features are not totally off the rails. 

A satellite level overview, in extremely simple layman's terms, of what each file, class and function are actually doing would help a lot. I think it would be very generous to myself to say that I could even identify the purpose of 5% of the code in this project, if you pointed at something specific and asked me, "What does this do?"

About one sentence each in plain English (suitable for a 5-year-old) for each file/class/function would probably clarify a lot already.

**Comment by RedBearAK** [comment]
@joshgoebel 

There is another aspect of the Option key special character entry in macOS that I would like to be able to mimic. It's possible with AutoHotKey already. But I don't know how difficult it would be to add it to the way `keyszer` processes nested shortcuts. You can actually do fairly complex actions with AHK after the initial combo, including sending some keys, and then still be waiting to trigger one of the second layer of combos or keys. 

Right now I don't think there is a way to run a command or send out some keystrokes immediately after the initial/outer layer of the nested shortcut is invoked, and then continue to process the second layer of combos/keys. 

An example of the way this would look is Option+Grave would place a visible Grave character so the user can see it, then the key pressed after that would send out a Backspace to erase the Grave character and replace it with the accented character. This is how it works in macOS. You can see the accent that is about to be applied to the next key. 

Or, the accent is actually placed in a highlight in macOS, which makes it even more intuitive that the next character will replace it. So the most precise way to mimic this might be to output the Grave, do Shift+Left, and then wait for the next key, which will then overwrite the highlighted accent character. This is easy to do with a macro, but only at the second level of nesting, so it's fairly pointless and would just be for show. 

```python
    C("Alt-Grave"): {                       # Dead Keys Accent Grave
        C("A"):                     [UC(0x0060),C("Shift-Left"),sleep(0.3),UC(0x00E0)],     # à Latin Small a with Grave
    },
```

In its current form, the Option-key schemes for both Windows and Linux don't attempt to do anything like this. So the user sees nothing in output until you press the second key in the dead keys sequence. It makes the dead keys usage less intuitive than it is on macOS, if you're not already familiar with the accent keys. 

If there is already a syntax that would perform an action for the initial nested shortcut and then continue on to process the inside layer(s) of nested shortcuts afterward, let me know.

**Comment by joshgoebel** [comment]
> Although I may be misjudging the difficulty due to lack of experience.

Perhaps.  I gave a step by step list,  you could start there and see how far you get... you could also go look at the code from before the pass thru was originally removed. (find the commit, etc)

I still highly recommend something like https://exercism.org/ and just focusing on learning more Python and as your skills grow then so will your comprehension.

**Comment by joshgoebel** [comment]
Closing in favor of #63 and #66.

> About one sentence each in plain English (suitable for a 5-year-old) for each file/class/function

Now that's a little young... in my mentoring I always use 10 years old as my canonical age for keeping things simple. :-)  I should add more comments for sure...
